# Curso Nivelador

## Jupyter Notebook

Jupyter Notebook es una aplicación web que nos permite crear y compartir documentos que contienen código Python, ecuaciones, visualizaciones y texto. Es una herramienta muy útil para desarrollar aplicaciones de Inteligencia Artificial, ya que nos permite ejecutar código Python en el navegador de forma interactiva.

Para ejecutar Jupyter Notebook, debemos abrir Anaconda Navigator y pulsar el botón "Launch" en la sección de Jupyter Notebook. Una vez abierto el navegador, podemos crear un nuevo cuaderno pulsando el botón "New" y seleccionando la opción "Python 3". Una vez creado el cuaderno, podemos ejecutar código Python en las celdas de código.

# Python

**Python** es un lenguaje de propósito general, interpretado, interactivo, orientado objetos y de alto nivel. Python se ha hecho muy popular debido a su sintaxis clara, la facilidad como lenguaje para crear *scripts* y la cantidad de librerías populares que permiten solventar casi cualquier problema con el que el programador se enfrente.




Otra de las ventajas de *Python* es que se encuentra presente en Windows, macOS y la mayoría de distribuciones de Linux. Para poder utilizar este lenguaje es necesario primero hacer una instalación previa, pero como hemos comentado previamente, al utilizar un *Jupiter Notebook* podremos ejecutar aquí código sin necesidad de esa configuración.

## Hola Mundo



In [ ]:
# First program
print("Hello World")

Aquí podemos ver nuestro primer programa en Python. Tan sencillo como escribir una línea de código para que se imprima en pantalla el mensaje que queramos.

<a name="analisis-lexico"></a>
## Análisis Léxico

El análisis léxico engloba a todos los tokens especiales que puede interpretar el interprete de Python.

En la [guía oficial](https://docs.python.org/3/reference/lexical_analysis.html) se puede ver en más profundidad estos términos.

### Identificadores y palabras reservadas



Los identificadores son nombres usados para identificar una variable, función, clase, módulo u otro objeto. Los identificadores pueden empezar con letras (A-Za-z), números (0-9) o una barra baja (_).

Python es un lenguaje que distingue entre mayúsculas y minúsculas, por lo que **Variable** y **variable** son dos identificadores diferentes en *Python*.






In [ ]:
# Identifiers

years_person = 10
# yearPerson = 10
# firstIdentifier = 10

second_identifier = "Hello"

third_identifier = { "hello": "world" }
_private_identifier = "Private"

if __name__ == "__main__":
  print("This is a magic method")
  print(years_person)
  print(third_identifier)

### Lineas e identación


Python no utiliza corchetes ({}) para indicar bloques de código dentro de las clases y funciones o control de flujo. En este lenguaje se utiliza la identación para indicar bloques, y se aplica sin ningún tipo de excepción.

El número de  espacios puede variar dependiendo del intérprete (se suele elegir 2 o 4) pero una vez elegido debe respetarse, por ello este ejemplo no produce ningún error:

In [ ]:
if True:
  print("Success")
else:
  print("Failed")

print("this is running anyway")

Mientras que si se cambian los espacios se producirá un error, como indicará el intérprete en este trozo de código:

In [ ]:
if True:
print("Sucess")
else:
  print("Failed")

Hay algunas excepciones como las listas o la palabra reservada `\`

### Comentarios

El símbolo de la almohadilla (#) representa un comentario en Python, toda cadena que precede a este símbolo en una misma linea es considerado un comentario en Python y es tratado así por el interprete. Además de esto contamos con tres comillas para comentarios multilinea.

In [ ]:
# This is a comment

test = "string" # You can place a comment after a line of code


'''
This is a multiline comment
You can write multiple statements between the quotes
'''


<a name="variables"></a>
## Variables y Tipos


Una variable es un espacio de memoria reservado dentro de una computadora para almacenar valores. En python la declaración se realiza al asignar un valor a una variable con el símbolo igual (=)


In [ ]:
# Variable asignment
pi_number = 3.14

print(pi_number)

pi_number = "hello world"
# Output variable
print(pi_number)

### Números

El primer tipo de dato en *Python* que vamos a ver son números. Los números son objetos creados cuando asignas un valor numérico a una variable.

*Python* soporta tres tipos de números:


*   **int**: Enteros con signo (Python 3 aúna int y long)
*   **float**: Números reales con coma flotante
*   **complex**: Números complejos


In [ ]:
# Number examples

# Integer
int_number = 10
print(int_number)
# Float
float_number = 12.40013243
print(float_number)
# Complex
complex_number = 3.14j
print(complex_number)


Algunas de las operaciones más importantes se pueden ver en su [documentación oficial](https://docs.python.org/3/library/stdtypes.html#numeric-types-int-float-complex)

### Cadenas



Las cadenas en Python se definen como una concatenación de caracteres representados entre comillas. Dentro de *Python* se pueden usar comillas simples ( ' ) o dobles ( " ) para representar cadenas.

In [ ]:
# String examples
name_person = 'Hello World'
char = "H"
str_2 = "Hello World again"

print(name_person)
print(char)
print(str_2)
print(name_person + "  ->  " + char)
print("==============")


name_person = "Other value"
print(name_person)
print("==============")

# Substrings

str1 = 'Hello World'
str2 = 'Hello World again'

print(str1[0]) # Prints first char
print(str1[0:5]) # Prints first char
print(str1[6:]) # Prints substring from second to sixth character
print(str1[:7])
print(str1[0:7])

Algunas operaciones se pueden ver en la [documentación de Python](https://docs.python.org/3/tutorial/introduction.html#strings)



Es posible dar formato al texto de diferentes formas, algunas incorporadas en las últimas versiones de python.

In [ ]:
# First form interpolation
interpolation1 = "Hello world I'm %s and I'm %d years old" % ("Javier", 32)
print(interpolation1)

# Second form interpolation
interpolation2 = "Hello world I'm {0} and I'm {1} years old".format("Javier", 32)
print(interpolation2)

# Third form interpolation (Python 3.6)
name = "Lucas"
age = 32
interpolation3 = f"Hello world I'm {name} and I'm {age} years old"
print(interpolation3)

Artículo sobre los distintos métodos de interpolación:

https://realpython.com/python-f-strings/


### Listas



Las listas son colecciones ordenadas de elementos que pueden no ser del mismo tipo. En python las listas se definen con valores separados por comas contenidos entre corchetes.

In [ ]:
list_example_strings = ["Hello", "World", "I'm", "John"]
print(list_example_strings)
list_example_mix = ["Hello", 2, "Number", 16]
print(list_example_mix)
list_example_mix_crazy = ["Hello", ["hello", "world"], "Number", 16]
print(list_example_mix_crazy)

#accessing values
element1 = list_example_strings[0]
print(element1)

# Error
#element_fail = list_example_strings[9]
#print(element_fail)

print("===========")

# Add element
first_list = ["Hello", "World"]
first_list.append("Again")
print(first_list)
first_list.insert(0, "****")
print(len(first_list))
print(first_list)


print("===========")

# Update element
updating_list = ["Hello", "World", 2017]
updating_list[2] = 2023
print(updating_list)

# Slicing
list_example = ["Hello", "World", "I'm", "John"]
print(list_example[1:])

# Simple indexing
simple_index = list_example[-1]
print(simple_index)

Para ampliar sobre operadores, accesos, optimización y sintáxis podéis acceder a su [documentación](https://docs.python.org/3/tutorial/introduction.html#lists)

### Tuplas


Las tuplas son secuencias inmutables de objetos en Python. Son muy parecidas a las listas con la diferencia que las listas es un conjunto de elementos variables mientras que las tuplas son inmutables. Para crear una tupla, se definen unos valores separados por comas contenidos entre paréntesis.

In [ ]:
# Tuple examples
tuple_example_mix = ("Hello", 2, "Number", 16)
print(tuple_example_mix)

tuple_not_parentesis = "hello", "world"
print(tuple_not_parentesis)

# tuples in functions
def get_tuple():
  return "Hello", "World"

hello, world = get_tuple()

print(f"{hello} {world}")
print(hello)
print(world)

Listas y Tuplas, artículo de refuerzo:

https://realpython.com/python-lists-tuples/


### Diccionario

Los diccionarios son una colección de valores sin orden que pueden intercambiarse y están indexados por una clave. Estas claves deben ser únicas, es decir, no puede haber repetición y mientras que el valor los valores puede mutar, las claves son inmutables.

In [ ]:
# Dictionary definition
empty_dict = {}
print(empty_dict)
string_dict = {"first": "Hello",
               "second": "World"}
print(f'{string_dict["first"]} {string_dict["second"]}')

# Accesing Values
string_dict = {"first": "Hello", "second": "World"}
ex_str = string_dict["first"]
print(ex_str)

# Adding values
dict_test = {"first": "Hello", "second": "World"}
dict_test["third"] = ["hola"]
print(dict_test)

# Updating element
dict_test = {"first": "Hi", "second": "World"}
print(f"We have this {dict_test}")
dict_test["first"] = "Hello"
print(f"We have this {dict_test}")

Para conocer los métodos y funciones de los diccionarios [podéis visitar la documentación oficial](https://docs.python.org/3/tutorial/datastructures.html?highlight=dictionary#dictionaries)


<a name="operadores"></a>
## Operadores



Los operadores son construcciones de Python que pueden manipular el valor de los operandos.

### Operadores Aritméticos



Los operadores aritméticos  modifican el valor de una variable.

In [ ]:
# Ejemplo de operador aritmético
a = 30
b = 50

# Addition
add = a + b
print(f"add: {add}")

# Subtraction
sub = a - b
print(f"sub: {sub}")

# Multiplication
mul = a * b
print(f"mul: {mul}")

# Division
div = a / b
print(f"div: {div}")

# Module
modd = b % a
print(f"modd: {modd}")

# Exponent
exp = a**b
print(f"exp: {exp}")

### Operadores de comparación

Comparan los valores de ambos lados de la operación y decide la relación entre ambos.

In [ ]:
# Ejemplos de operadores de comparación
a = 30
b = 50

# Equality
eq = a == b
print(f"eq: {eq}")

# Inequality
ieq = a != b
print(f"ieq: {ieq}")

# Greater
gt = a > b
print(f"gt: {gt}")

# Less
ls = a < b
print(f"ls: {ls}")

# Greater or equal
gteq = a >= b
print(f"gteq: {gteq}")

# Less or equal
lseq = a <= b
print(f"lseq: {lseq}")


### Operadores de asignación

Hemos visto estos operadores reiteradamente, pero vamos a explicar ahora las propiedades de algunos de ellos.

In [ ]:
# Ejemplos de operadores de asignación
a = 30
b = 50

# Assignment (=)
c = a + b
print(c)

# Addition with assignment
c = 60
c += a
# c = c + a
print(c)

# Subtraction with assignment
c = 60
c -= a
# c = c - a
print(c)

# Multiplication with assignment
c = 60
c *= a
# c = c * a
print(c)

# Division with assignment
c = 60
c /= a
# c = c / a
print(c)

### Operadores lógicos, de identidad y pertenencia

Vamos a agrupar tres tipos de operadores que son utilizados detro de lso controles de flujo que veremos a continuación, todos devuelven  un tipo lógico (true or false) cuando se realiza la operación.

In [ ]:
a = 30
b = 50
c = [20, 30, 70, 80]
at = True
bf = False

# AND operator
print(at and bf)
print(a > b and a < b)

print("------------")

# OR operator
print(at or bf)

print("------------")

# NOT operator
print(not at)

print("------------")

# Membership
print(a in c)

print("------------")

# Not Membership
print(a not in c)

print("------------")

# Identity
c = a
print(c)
print(a)
print(a is c)
print(a is b)

# Not identity
print(a is not c)
print(a is not b)


<a name="control-flujo"></a>
## Control de flujo



En el nivel más elemental, un programa no es más que una ejecución secuencial de instrucciones. Esta ejecución tiene un flujo predefinido que el desarrollador querrá controlar con diversas técnicas.

Para conocer más sobre el control de flujo el [tutorial que proporciona Python](https://docs.python.org/3/tutorial/controlflow.html) tiene todos los operadores necesarios.

### Decisión condicional

In [ ]:
# If example
counter = 1
if(counter == 5): #conditional test
  print("We've reached 5")
  print("")
  print("")
print("i'm always running")

# If..else example
if(counter == 6):
  print("We've reached 6")
  print("Hello")
else:
  print("We've got something else")

#counter += 1
# Nested if
if(counter == 1):
  print("We've reached 1")
elif(counter == 2):
  print("We've reached 2")
else:
  print("We've got more than that")


### Bucle "for"

In [ ]:
# For loop example
elements = [1,2,3,4,5,6,7,8]

for element in elements:
  print(element)
  print("Hola")

for letter in "Hello World":
  print(letter)


### Bucle "while"

In [ ]:
# While loop
counter = 1

while (counter < 9):
  print(counter)
  counter += 1

In [ ]:
# While loop infinite DO NOT EXECUTE

counter = 1

while (counter != 2):
  print(counter)
  counter += 2


<a name="funciones"></a>
## Funciones

Una función es un bloque organizado de código reutilizable. Las funciones otorgan  modularidad a una aplicación y un nivel de reusabilidad. Hay varios consejos que debe seguir una función:



*   Las funciones deben ser lo más nucleares posibles: Una función debe realizar una sola tarea con un objetivo claro
*   Cualquier código que se repita en tu programa es candidato de convertirse en una función.
*   Las funciones deben ser descriptivas en su funcionamiento: El nombre de la función debe  dar una descripción aproximada de su funcionamiento.



In [ ]:
def function_example(parameter):
  """DocSttring"""
  print(parameter)
  return

In [ ]:
function_example("Hola Mundo")

Se puede realizar tantas llamadas como sea necesario, cambiando el valor de los parámetros cuando se quiera.

In [ ]:
function_example("Que tal estás")
function_example("Que tal asdfasdf")
function_example("Que tal estás")

### Parámetros de una función

En Python pueden utilizarse cuatro tipos de argumentos en una función:





**Argumentos obligatorios**

Los argumentos obligatorios son aquellos que se pasan a una función en el orden correcto y con el número exacto de variables para argumentos.

In [ ]:
def required_args(first, second, third):
  print(first)
  print(second)
  print(third)

required_args("first", "2", "third")

**Argumentos por defecto**

Un argumento por defecto es un argumento que asume su valor por defecto, estos argumentos siempre van después de los argumentos por clave y pueden ser omitidos en la llamada a la función.

In [ ]:
def default_args(first, second, third="third", fourht="4"):
  print(first)
  print(second)
  print(third)
  print(fourht)
  print("========")

# You can call the third value
default_args("first", "second", "3")

# You can ommit the value
default_args("first", "2")

# You can change the default value
default_args("first", "second", fourht="other")

**Return**

La palabra reservada **return** sirve para terminar una función y devolver una expresión a la linea de código que la ha llamado. Una función sin **return** es lo mismo que si devuelve None.

Puede devolverse cualquier valor, y como mencionamos en el curso anterior, pueden **devolverse múltiple valores mediante tuplas.**

In [ ]:
def simple_sum(a, b):
  return a + b

result = simple_sum(4, 2)
simple_sum(4, 8)
print(result)


def increment_two_values(a, b):
  a += 1
  b += 1
  return a, b

inc_1, inc_2 = increment_two_values(3, 5)
print(f"{inc_1} and {inc_2}")


<a name="orientacion-objetos"></a>
## Python Orientado a Objetos







Podemos decir que en la Orientación a Objetos estructuramos **variables** y **funciones** en unidades lógicas, lo que llamaremos objetos. Estas **variables** son lo que hemos llamado **atributos** y las **funciones** serán los **métodos**.

Un ejemplo de objeto sería un coche. Un coche tiene **atributos** importantes como el color, el modelo, el año y el tipo de gasolina. Además podemos interactuar con el objeto mediante sus **métodos** como arrancar, o acelerar.

Aunque sea sorprendente, ya hemos usado múltiples objetos en Python, por ejemplo las listas. Las listas tienen **métodos** como append(), clear() o copy() y se organizan en unidades lógicas con una función.



### Clases

Son "prototipos" definidos por el programador que sirven como modelo para los atributos y los métodos de un objeto. Se podría hacer una analogía con las recetas de un plato. Con las clases podemos definir qué valores tendrá el objeto y qué podrá hacer.

Para crear una clase solo necesitamos declarar la palabra reservada *class* y definir la clase interna:



```
class NewClass(object):
    # Documentation
    class_body
```

Así podremos ver un ejemplo de clase completo con un coche:


In [ ]:
class Car(object):
  'Common class for a car object'
  total_cars = 0 #secuencial number

  # Declaration
  def __init__(self, model, color, id="2342FSD", height=2):
    self.model = model
    self.color = color
    self.id = id
    self.height = height
    Car.total_cars += 1

  def get_model(self):
    return self.model

  def print_model(self):
    print(f"The car is {self.model}")

  def print_total_cars(self):
    print(f"Total amount of cars: {Car.total_cars}")

  def print_color(self):
    print(f"The color of the car is {self.color}")

  def start(self):
    print("Engine running")


Aquí podemos ver tres elementos importantes:



*   La variable *total_cars* es una variable compartida por todas las instancias de la clase, puede acceder estáticamente a través de una instancia *ClassName.variable*, en este caso *Car.serial_number*
*   El primer método *\_\_init__()* es un método especial reservado a la inicialización del objeto, llamado **constructor**. Este es llamado cuando se instancia un objeto.
*    Se declaran los atributos y los métodos, que son proipos de cada instancia. Para ello se usa la palabra reservada *self* tanto como primer argumento de un método como para referenciar a un atributo.



### Objetos

Un objeto es una instancia de una estructura de datos definida en una clase. Un objeto se compone tanto de **atributos** de clase como de métodos.

Para crear una instancia de una clase, solo hay que asignar a una variable la llamada al constructor de una clase y pasar los argumentos que acepta el constructor *\_\_init__*

En Python la instanciación de un objeto sería de la siguiente forma:



```
inst_object = NewClass(attribute)
```



In [ ]:
# Objetc instantiation

opel_car = Car("opel", "red", "67458IUK")

default_car = Car("test", "red")

mercedes_car = Car("mercedes", "blue", "2342IOS")

tesla_car = Car("tesla", "grey", "1234ABC")

patata_car = Car("patta", "grey", "asdfasf")


#fail_car = Car()

#### Métodos y atributos

Una vez instanciado el objeto, podremos acceder a sus atributos y métodos, para ello usaremos la notación de python (que ya conocemos bastante bien) con la llamada a la variable, seguida de un punto y el nombre del atribtuo o método.



```
isnt_object.attribute
inst_object.method()
```

Así vemos como acceder a los atributos de nuestros objetos ya instanciados:


In [ ]:
# method call
opel_car.print_model()

# attribute_call
print(mercedes_car.color)

print(tesla_car.id)

En las llamadas a métodos, pese a que tenga como primer argumento *self* no es necesario declararlo, ya que como comentamos es una palabra reservada de Python para declarar una función como un método de clase.

Por otro lado, podremos modificar atributos de clase normalmente:

In [ ]:
opel_car.color = "green"
opel_car.print_color()


## I/O

**Print**



In [ ]:
print("Hello World")
test = "Other"
print(test)

**Input**



In [ ]:
name = input("Say your name:")

print(f"Hello, {name}")

**Open**

La función open permite abrir determinados archivos, su sintaxis es la siguiente:



```
> f = open('workfile', [flag])
```









**File**

La función **open** nos devuelve un objeto de tipo **file**. Con este objeto podremos realizar las siguientes operaciones:



*   **file.closed** -> Nos devuelve si el archivo está cerrado.
*   **file.mode** -> Nos indica el tipo de acceso.
*   **file.name** -> Devuelve el nombre del archivo.
*   **file.close()** -> Función que cierra y el archivo y serializa los cambios.



**CSV**

Uno de los formatos más comúnes dentro del mundo de Data Science es organizar los datos en CSV delimitados, para ello hay que usar la función **open** y hacer un cast a **csvfile** y llamar al lector de csv, como en el siguiente ejemplo:



```
import csv

with open('data.csv') as csvfile:
        reader = csv.DictReader(csvfile, delimiter=';')
        for row in reader:
```



## Excepciones

Los tipos de excepciones que podemos encontrar son enormes y están recogidos en la [documentación oficial de Python](https://docs.python.org/3.7/library/exceptions.html).

La sintaxis para el tratamiento de errores es la siguiente:



```
try:
    Aquí va el código que queremos controlar, siempre identado...
    .......
    .......
except Exception1 as e:
    Aquí tratar la excepción 1 que hemos renombrado a "e"
except Exception2:
    Puede ocurrir otro tipo de excepción
    .........
    .........
else:
    Este bloque se ejecuta si no ha habido errores.
    
```




In [ ]:
dict_test = {"hello": "hello", "world": "world"}

try:
  test = dict_test["other"]
except KeyError as e:
  print(e)
  test = "Not found"

print(f"We reach this part of the program with {test}")

<a name="gestion-modulos"></a>
## Gestión de módulos

Un módulo es un fichero Python donde se pueden definir funciones, clases y variables que puedes referenciar y añadir a otro código.

Este código puede importarse usando la palabra reservada **import** o **from** con el siguiente formato:



```
import module1, module2.....
from module1 import [reference]
```



In [ ]:
# Import module
import time

# call module by callint it's name
seconds = time.time()

print(seconds)

Por otro lado, con la notación **from ... import** podemos importar atributos específicos de un módulo al *namespace* actual.

In [ ]:
from time import time

seconds = time()

print(seconds)

[Aquí](https://docs.python.org/3/tutorial/modules.html) podéis encontrar más información acerca de módulos.

## Librerías de Tratamiento de Datos



*   [Numpy](https://www.numpy.org/) -> Librería destinada a la manipulación de matrices, especialmente dentro de operaciones matemáticas.
*   [SciPy](https://www.scipy.org/) -> Otra librería de Python orientada a computación técnica y científica.
*   [Matplotlib](https://matplotlib.org/) -> Muy útil para visualización de datos.
*   [Pandas](https://pandas.pydata.org/) -> Especializada en manipulación de datos y posterior análisis.
*   [Scikit-Learn](https://scikit-learn.org/stable/) -> Paquete de Python diseñado para dar acceso a algoritmos de Aprendizaje Automático bien conocidos dentro del código Python, a través de una API limpia y bien pensada. Ha sido construido por cientos de colaboradores de todo el mundo y se usa en la industria y el mundo académico.
Scikit-Learn esá basada en las librerías **NumPy** (Numerical Python) y **SciPy** (Scientific Python), que permiten la computación numérica y científica eficiente en Python.




### Tratamiento de datos

Vamos a ver como procesar en Colab csv, utilizando una de las librerías antes mencionadas.

Para ello vamos a usar el dataset de [abalones](https://es.wikipedia.org/wiki/Haliotis) de la [uci](https://archive.ics.uci.edu/ml/datasets/abalone).

Para ello, descargar el archivo **Abalones** de la carpeta principal.

Para importar los archivos, vamos a utilizar la librería `google.colab`, que permite controlar aspectos del cuaderno, para forzar la carga de archivos. Posteriormente con la librería `io` serializaremos estos datos.

En un código ejecutado en una máquina este paso no sería necesario, podríamos usar `pd.read_csv` en la carpeta donde se encontrase nuestro archivo.

In [ ]:
import pandas as pd
from google.colab import files
import io

uploaded = files.upload()
abalone_train = pd.read_csv(io.BytesIO(uploaded['abalone_train.csv']), names=["Length", "Diameter", "Height", "Whole weight", "Shucked weight",
           "Viscera weight", "Shell weight", "Age"])

abalone_train.head()

# Ejemplo de Red Neuronal en Python

## Introducción

En este ejemplo veremos un modelo de Red Neuronal usando el dataset de Dígitos de MNIST. El objetivo es familiarizarnos con el código en Python, no es necesario saber lo que está pasando pero haremos una introducción para añadir contexto.

Algunas de las figuras que contienen el dataset serían las siguientes.

![MNist](https://upload.wikimedia.org/wikipedia/commons/2/27/MnistExamples.png)

El objetivo de este modelo consiste en encontrar pesos que nos asegure la evidencia de la existencia de cada uno de los dígitos, sin usar la información espacial del pixel.

![NeuralNetworkMNISTLow](https://user-images.githubusercontent.com/16117276/221560697-ff9c2629-f394-4a7a-9ec2-f8733a06df03.gif)

*Explicación de [3Blue1Brown](https://youtu.be/aircAruvnKk)*


Es por ello que para el siguiente ejemplo usaremos Keras para:


1. Extraer datos para el entrenamiento y la evaluación.
2. Construir una red neuronal que clasifique imágenes.
3. Añadir capas ocultas para su entrenamiento.
4. Entrenar esa red neuronal.
5. Evaluar la precisión del modelo.

El código que vamos a utilizar es el siguiente:

In [1]:
import tensorflow as tf
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10)
])

loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

model.fit(x_train, y_train, epochs=5)

model.evaluate(x_test,  y_test, verbose=2)

11490434/11490434 [==============================] - 1s 0us/step
Epoch 1/5
  74/1875 [>.............................] - ETA: 1s - loss: 1.0971 - accuracy: 0.6744  

2023-08-27 21:49:02.682866: W tensorflow/tsl/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz


1875/1875 [==============================] - 1s 674us/step - loss: 0.2979 - accuracy: 0.9122
Epoch 2/5
1875/1875 [==============================] - 1s 670us/step - loss: 0.1448 - accuracy: 0.9575
Epoch 3/5
1875/1875 [==============================] - 1s 677us/step - loss: 0.1100 - accuracy: 0.9671
Epoch 4/5
1875/1875 [==============================] - 1s 671us/step - loss: 0.0886 - accuracy: 0.9723
Epoch 5/5
1875/1875 [==============================] - 1s 650us/step - loss: 0.0758 - accuracy: 0.9762
313/313 - 0s - loss: 0.0752 - accuracy: 0.9770 - 173ms/epoch - 552us/step


[0.07520347833633423, 0.9769999980926514]

## Desarrollo del modelo

Vamos ahora a ver el modelo en más profundidad, analizando la sintaxis de Python que usamos en el ejemplo.

### Importar dependencias

Lo primero sería importar tensorflow y preparar el dataset de MNIST, para ello vamos a acceder al módulo `mnist`. Tensorflow tiene una gran variedad de datos de catalogo para entrenar nuestros modelos, puedes consultarlos [aquí](https://www.tensorflow.org/datasets/catalog/overview). En este ejemplo vamos a usar la librería de **tensorflow** y el dataset de **mnist**, que asignaremos a una nueva variable.

Para ello, vamos a hacer uso de lo siguiente:

- [Análisis lexico](#analisis-lexico)
- [Gestión de modulos](#gestion-modulos)

In [ ]:
import tensorflow as tf
mnist = tf.keras.datasets.mnist


### Establecer los datos de entrenamiento

El siguiente paso será establecer los datos para entrenamiento y test, en este caso convertiremos los datos de enteros a numeros con coma flotante, para ello vamos a llamar a una **función** dentro del fichero del paquete mnist, que nos devolverán unas **tuplas** que desempaquetaremos en **cuatro variables**, cada una formada por una **lista**.

- [Variables, Listas y Tuplas](#variables)
- [Funciones](#funciones)
- [Operadores](#operadores)

In [ ]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

### Crear el modelo

Ahora constuiremos el modelo con la API ```tf.keras.Sequiential```, como podemos observar, podemos declarar las capas en variables o directamente instanciarlas en la propia declaración del modelo. Elegiremos un optimizador y una función de perdida. El modelo será un **objeto** que crearemos instanciando una **clase**.

- [Orientación a objetos](#orientacion-objetos)

In [ ]:
# Se pueden asignar a variables y después pasarlas como argumentos
l0 = tf.keras.layers.Flatten(input_shape=(28, 28))

# O directamente instanciarlas como argumentos
model = tf.keras.models.Sequential([
  l0,
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dropout(0.2),
  tf.keras.layers.Dense(10)
])

Vamos a ver la información de las capas antes del entrenamiento, para ello vamos a crear nuestra propia [funcion](#funciones), para ello vamos a iterar sobre las distintas capas del modelo.

- [Control del flujo](#control-flujo)

In [ ]:
def display_layers(layers):
  for index, layer in enumerate(layers):
    print(f"-------------- {layer.name} --------------")
    print(layer.weights)
    print(f"------------------------------")

display_layers(model.layers)

También se puede ver un resumen del modelo de forma sencilla con este **método** del **objeto** model.

In [ ]:
model.summary()

### Función de perdida

La función de perdida `losses.SparseCategoricalCrossentropy` coge un vector de logits y un Inidce `True`y devuelve una perdida escalar por cada ejemplo. En este caso estamos instanciando otro **objeto** y simplemente le indicamos que usaremos **logits** (probabilidades) como entrada.

In [ ]:
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

### Compilar el modelo

Vamos a configurar el modelo para su entrenamiento, como vimos en el ejemplo anterior vamos a hacer uso de una **función de optimización** y una **función de pérdida** para entrenar el modelo.

Además, vamos a introducir el concepto de **métricas**, básicamente son funciónes similares a la función de perdida, que permiten evaluar la precisión del modelo, pero a diferencia de la función de perdida, no es usada en el entrenamiento.

Para ello otra vez llamarémos a um **método** del objeto **model** y pasaremos los atributos necesarios.


In [ ]:
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

### Entrenar el modelo

Vamos a entrenar el modelo con los datos de entrenamiento llamando al método `fit`.

Símplemente vamos a coger como parámetros las **features** (`x_train`) las **labels** (`y_train`), los ciclos de cada entrenamiento (`epochs`) y con ello nos va a devolver un objeto `history` con el que medir el entrenamiento.

Ahora llamaremos a otro de los **métodos** del objeto para entrenar, pasando las variables de **features** y **labels**.

In [ ]:
history = model.fit(x_train, y_train, epochs=5)

Así, los pesos de las capas quedarían así:

In [ ]:
display_layers(model.layers)



### Mostrar las estadísticas del entrenamiento.

El método `fit` devuelve un objeto `history`. Podemos usar este objeto para mostrar una gráfica con la variación de la **función de perdida** entre cada *epoch*. Una perdida grande significa que la predicción está muy lejos del resultado.

Vamos a usar [Matplotlib](https://matplotlib.org/) para visualizar el resultado. Como podemos ver, el modelo mejora rápidamente para luego frenarse al final de las ejecuciones.

Para ello volveremos a **importar un módulo**, y de ahí llamar a las **variables** y **funciones** que necesitemos.


In [ ]:
import matplotlib.pyplot as plt
plt.xlabel('Epoch Number')
plt.ylabel("Loss Magnitude")
plt.plot(history.history['loss'])

### Mostrar el desempeño del modelo

El **método** `evaluate` del **objeto** `model` comprueba el desempeño del modelo. Vamos ahora a usar un **flujo condicional** para comprobar si la precisión es de más de un 90%, y si es así imprimir un mensaje en pantalla.

- [Control del flujo](#control-flujo)

In [ ]:
evaluation = model.evaluate(x_test,  y_test, verbose=2)
model_accuracy = evaluation[1]

if(model_accuracy > 0.9):
  print("Nuestro modelo acierta más del 90%")
else:
  print("Nuestro modelo tiene una precisión por debajo del 90%, repetir el proceso")

Ahora el modelo está entrenado para tener una probabilidad del 97%